src/minimization.cpp

```diff
@@ -0,0 +1,289 @@
+}
+
+int getPortIndex(TreeNode::Id id, TreeNode &node) {
+  for (size_t i = 0; i < node.succ.size(); ++i) {
+    if (node.succ[i] == id)
+      return i;
+  }
+  for (size_t i = 0; i < node.pred.size(); ++i) {
+    if (node.pred[i] == id)
+      return i;
+  }
+    return 0;
+}
+
+void edgesByLayers(Net &net,
+     std::vector<TreeNode::Id> &vec,
+     TreeNode *node,
+     std::vector<Edge> &edges) {
+  for (TreeNode::Id &id: vec) {
+    TreeNode *connectedNode = net.getNode(id);
+    if (connectedNode->layer > node->layer)
+      continue;
+    edges.emplace_back(connectedNode, node);
+  }
+}
+
+std::vector<std::vector<Edge>> getNetEdges(Net &net,
+                               std::vector<std::vector<TreeNode::Id>> &tempNodesByLayer) {
+  std::vector<std::vector<Edge>> netEdges;
+  for (size_t i = 1; i < tempNodesByLayer.size(); ++i) {
+    std::vector<Edge> edges;
+    for (size_t j = 0; j < tempNodesByLayer[i].size(); ++j) {
+      TreeNode *node = net.getNode(tempNodesByLayer[i][j]);
+      edgesByLayers(net, node->pred, node, edges);
+      edgesByLayers(net, node->succ, node, edges);
+    }
+    netEdges.push_back(edges);
+  }
+  return netEdges;
+}
+
+bool lexicographicSortCondition(const Edge &edge1, const Edge &edge2) {
+  if (edge1.first->number == edge2.first->number) {
+    return edge1.second->number < edge2.second->number;
+  } else {
+    return edge1.first->number < edge2.first->number;
+  }
+}
+
+int nearestPow2(int x) {
+  --x;
+  x |= x >> 1;
+  x |= x >> 2;
+  x |= x >> 4;
+  x |= x >> 8;
+  x |= x >> 16;
+  return ++x;
+}
+
+int AdditionalNetFeatures::crossCounting(AdditionalNetFeatures &features) {
+  int crossCount = 0;
+  for (size_t i = 0; i < netEdges.size(); ++i) {
+    std::sort(netEdges[i].begin(), netEdges[i].end(), lexicographicSortCondition);
+
+    const int numLeaves = nearestPow2(netEdges[i].size());
+    const int firstLeafIndex = numLeaves - 1;
+    const int treeSize = numLeaves * 2 - 1;
+
+    features.accTree.clear();
+    features.accTree.resize(treeSize, 0);
+
+    for (size_t k = 0; k < netEdges[i].size(); k++) {
+      int index = netEdges[i][k].second -> number + firstLeafIndex;
+      ++features.accTree[index];
+      while (index > 0) {
+        if (index % 2)
+          crossCount += features.accTree[index + 1];
+        index = (index - 1) / 2;
+        ++features.accTree[index];
+      }
+    }
+  }
+  return crossCount;
+}
+
+void getBarycentricValueForPorts(Net &net,
+     std::vector<TreeNode::Id> &vec,
+     bool forwardLayerSweep,
+     std::vector<std::pair<float, TreeNode::Id>> &barycentricValueForPorts) {
+  for (TreeNode::Id &id: vec) {
+    TreeNode *node = net.getNode(id);
+    float bValue;
+    if (forwardLayerSweep) {
+      bValue = forwardRankDefinition(*node, node->number, getPortIndex(node->id, *node)) /
+               float(node->succ.size() + node->pred.size());
+    } else {
+      bValue = backwardRankDefinition(*node, node->number, getPortIndex(node->id, *node)) /
```

Comment:
What `getPortIndex(node->id, *node)` means? Why we are searching for self-loops?